cryosparc/api.py

```diff
@@ -0,0 +1,409 @@
+
+    @contextmanager
+    def _request(self, method: str, url: str, *args, **kwargs) -> Iterator[httpx.Response]:
+        attempt = 0
+        max_attempts = 3
+        while attempt < max_attempts:
+            attempt += 1
+            try:
+                request = self._client.request(method, url, *args, **kwargs)
+                yield request
+                return
+            except httpx.TransportError as err:
+                warnings.warn(
+                    f"*** API client {method.upper()} {url} failed due to {err} (attempt {attempt}/{max_attempts})"
+                )
+                if attempt == max_attempts:
+                    raise
+
+    @contextmanager
+    def _request_stream(self, method: str, url: str, *args, **kwargs) -> Iterator[httpx.Response]:
+        attempt = 0
+        max_attempts = 3
+        while attempt < max_attempts:
+            attempt += 1
+            try:
+                with self._client.stream(method, url, *args, **kwargs) as response:
+                    yield response
+                    return
+            except httpx.TransportError as err:
+                warnings.warn(
+                    f"*** API client {method.upper()} {url} stream failed due to {err} (attempt {attempt}/{max_attempts})"
+                )
+                if attempt == max_attempts:
+                    raise
+
+        raise RuntimeError(f"[API] Could not complete request {method.upper()} {url}")
+
+    def _call(self, _method: str, _path: str, _schema, *args, **kwargs):
+        """Meta-call method that runs whenever a named function is called on
+        this namespace"""
+        responses_schema = _schema.get("responses", {})
+        _path, req = self._construct_request(_path, _schema, *args, **kwargs)
+
+        # Submit a response and get a generator for it. Resulting res_gen should
+        # only yield a single item. Cleaned up at the end of this function call.
+        content_schema = responses_schema.get("201", {}).get("content", {})
+        stream_mime_type = registry.first_streamable_mime(content_schema.keys())
+
+        try:
+            ctx = (
+                self._request(_method, _path, **req)
+                if stream_mime_type is None
+                else self._request_stream(_method, _path, **req)
+            )
+            with ctx as res:
+                return self._handle_response(_schema, res)
+        except httpx.HTTPStatusError as err:
+            raise APIError("received error response", res=err.response)
+
+
+class APIClient(APINamespace):
+    """Root API namespace interface for an OpenAPI server"""
+
+    _namespace_class = APINamespace
+
+    def __init__(
+        self,
+        base_url: Optional[str] = None,
+        *,
+        auth: Optional[Auth] = None,  # token or email/password
+        timeout: float = 300,
+        http_client: Optional[httpx.Client] = None,
```

Comment:
New signature, you must now specify `base_url` instead of separate `host` and `port` arguments. This will make it easier to create a client that calls a production CryoSPARC instance, e.g., `APIClient("https://cryosparc.example.com/api")`